### PR TITLE
pregel: changeable shardkeyattribute (devel)

### PR DESCRIPTION
**Pull request by ObiWahn**
## Scope & Purpose

Allow to use a different sharding key than `vertex`

- [x] Bug-Fix for devel
- [x] Bug-Fix for 3.4 -- https://github.com/arangodb/arangodb/pull/9157 

#### Related Information

- [x] There is an internal planning ticket: https://github.com/arangodb/team-orco-planning/issues/135

### Testing & Verification

- [x] Added new **integration tests**  tests/js/common/shell/shell-pregel-start.js

### Documentation

- [x] Added a *Changelog Entry* 
- [ ] Added entry to *Release Notes* 
- [x] Added a new section in the *Manual* 
- [ ] Added a new section in the *http API* 
- [ ] Added *Swagger examples* for the http API  

### CI
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/4756/

### CODE REVIEW

File: arangod/Pregel/PregelFeature.cpp [hunk 1]
```diff
@@ -73,6 +73,35 @@ struct NonDeleter {
 
 std::shared_ptr<arangodb::pregel::PregelFeature> instance;
 
+static std::string const shardKeyAttributeKey = "shardKeyAttribute";
```
Comment (jsteemann) [inline arangod/Pregel/PregelFeature.cpp]: Why is all this necessary if `shardKeyAttribute` is an optional attribute?

File: arangod/Pregel/PregelFeature.cpp [hunk 2]
```diff
@@ -162,10 +197,15 @@ std::pair<Result, uint64_t> PregelFeature::startExecution(
 
         if (!coll->isSmart()) {
           std::vector<std::string> eKeys = coll->shardKeys();
-          if (eKeys.size() != 1 || eKeys[0] != "vertex") {
+          if (eKeys.size() != 1 ||  params.get(shardKeyAttributeKey).compareString(eKeys[0])) {
```
Comment (jsteemann) [inline arangod/Pregel/PregelFeature.cpp]: I think changes would only be required here:
1) get shardKeyAttribute value from job data if specified. If not specified, use "vertex"
2) compare actual shard key with shardKeyAttribute devised in 1)

File: tests/js/common/shell/shell-pregel-start.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  'use strict';
+  var pid;
+  return {
+
+    ////////////////////////////////////////////////////////////////////////////////
+    /// @brief set up
+    ////////////////////////////////////////////////////////////////////////////////
+
+    setUp: function () {
+        db._create(vColl, { numberOfShards : 3 } )
```
Comment (jsteemann) [inline tests/js/common/shell/shell-pregel-start.js]: indentation by 4 here, and by 2 below...

File: tests/js/common/shell/shell-pregel-start.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
+/*jshint globalstrict:false, strict:false */
```
Comment (jsteemann) [inline tests/js/common/shell/shell-pregel-start.js]: I think this file should be named `-cluster.js`, as it is supposed to be executed only in cluster mode, right?

File: tests/js/common/shell/shell-pregel-start.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    setUp: function () {
+        db._create(vColl, { numberOfShards : 3 } )
+    },
+
+    ////////////////////////////////////////////////////////////////////////////////
+    /// @brief tear down
+    ////////////////////////////////////////////////////////////////////////////////
+
+    tearDown: function () {
+        pregel.cancel(pid); // delete contents
```
Comment (jsteemann) [inline tests/js/common/shell/shell-pregel-start.js]: will this work if `pid` is not set?

File: tests/js/common/shell/shell-pregel-start.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+// /
+// / Copyright holder is ArangoDB GmbH, Cologne, Germany
+// /
+// / @author Jan Chrisoph Uhde
+// / @author Copyright 2019, ArangoDB GmbH, Cologne, Germany
+// //////////////////////////////////////////////////////////////////////////////
+
+var jsunity = require("jsunity");
+var db = require("@arangodb").db;
+var graph_module = require("@arangodb/general-graph");
```
Comment (jsteemann) [inline tests/js/common/shell/shell-pregel-start.js]: not used

File: tests/js/common/shell/shell-pregel-start.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+var internal = require("internal");
+let pregel = require("@arangodb/pregel");
+
+const graphName = "UnitTest_pregel";
```
Comment (jsteemann) [inline tests/js/common/shell/shell-pregel-start.js]: not used

File: tests/js/common/shell/shell-pregel-start.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+// /
+// / Unless required by applicable law or agreed to in writing, software
+// / distributed under the License is distributed on an "AS IS" BASIS,
+// / WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// / See the License for the specific language governing permissions and
+// / limitations under the License.
+// /
+// / Copyright holder is ArangoDB GmbH, Cologne, Germany
+// /
+// / @author Jan Chrisoph Uhde
```
Comment (jsteemann) [inline tests/js/common/shell/shell-pregel-start.js]: what is the correct spelling of your name? :smile:

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/6301/

**Comment by ObiWahn** [comment]
https://github.com/arangodb/docs/pull/69

**Pull request merged**